netbench/netbench-test-player/src/status.rs

```diff
@@ -0,0 +1,184 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+use serde::{Deserialize, Serialize};
+
+use std::{
+    net::SocketAddr,
+    sync::{
+        atomic::{AtomicU8, Ordering},
+        Arc,
+    },
+    time::Duration,
+};
+
+use num_enum::{IntoPrimitive, TryFromPrimitive};
+
+use tokio::{
+    io::{AsyncReadExt, AsyncWriteExt},
+    net::{TcpListener, TcpStream},
+    time::{sleep, timeout},
+};
+
+use tracing::{info, instrument};
+
+/// The status of an endpoint; used to coordinate actions.
+#[derive(
+    Debug, Clone, Copy, Serialize, Deserialize, PartialEq, IntoPrimitive, TryFromPrimitive,
+)]
+#[repr(u8)]
+pub enum Status {
+    /// Still preparing the environment for the test.
+    ///
+    /// This may include installing dependencies, building drivers,
+    /// ect. endpoints may spend a minutes in this state.
+    NotReady,
+    /// Ready to begin running the running the collector and driver,
+    /// but have not started yet.
+    Ready,
+    /// Running the collector and driver.
+    Running,
+    /// Finished, the collector and driver are no longer running.
+    Finished,
+}
+
+#[derive(Debug, Clone)]
+pub struct StatusTracker {
+    // this endpoint's current state as an atomicU8 for safe sharing between threads
+    pub current_state: Arc<AtomicU8>,
+    pub remote_status_server: SocketAddr,
+    pub local_status_server: SocketAddr,
+}
+
+impl StatusTracker {
+    pub fn new(remote_status_server: SocketAddr, local_status_server: SocketAddr) -> Self {
+        Self {
+            current_state: Arc::new(AtomicU8::new(Status::NotReady.into())),
+            local_status_server,
+            remote_status_server,
+        }
+    }
+
+    /// Update our own state
+    #[instrument]
+    pub fn store(&mut self, value: Status) {
+        self.current_state.store(value.into(), Ordering::Relaxed)
+    }
+
+    /// Query peer for state; If connection fails return `assume_on_no_response`
+    #[instrument]
+    async fn get_status(remote_status_server: SocketAddr, assume_on_no_response: Status) -> Status {
+        let mut stream = match TcpStream::connect(remote_status_server).await {
+            Ok(stream) => stream,
+            Err(_) => return assume_on_no_response,
+        };
+        let mut buffer = Vec::new();
+        let peer_status = match stream.read_to_end(&mut buffer).await {
+            Ok(_) => serde_json::from_slice(&buffer).expect("Failed to parse peer's status"),
+            Err(_) => assume_on_no_response,
+        };
+        drop(stream);
+        peer_status
+    }
+
+    /// A Task that waits for peer to be in a particular state
+    #[instrument]
+    pub async fn wait_for_peer(
```

Comment:
I still think I'd prefer the strategy of "pushing state updates" rather than "pulling state updates"

This would allow you to get rid of the polling mechanism. It wouldn't be necessary to keep checking for updates, you'd simply await the TCP stream, and only wakeup when there is data to be read. You can use tokio synchronization primitizes like `Notify` or `Oneshot` to coordinate between the state tracker (which actually reads from the socket) and the application space with is waiting for that data.

I appreciate the commitment to error handling, but I think we should feel free to be less delicate. If something goes wrong, we don't really have a way of recovering. This test player was created with a single purpose, to run the test, and if it can't do that it should die.

In this scenario I would actually push for `wait_for_peer(status: State)` to be the main public api. Hid the implementation details (timeouts) inside, so that
- we will wait up to 3? minutes for the server to be ready, otherwise we panic and fail
- we will wait up to UPPER_TIME_LIMIT (10 minutes?) for the client to finish, otherwise we fail.